### PR TITLE
Grants route

### PR DESCRIPTION
**Pull request by kurtwheeler**
## Issue Number

#60 

## Purpose/Implementation Notes

I think the grants route is almost done. I just need to add the `grant/id/users` and `grants/id/organizations` sub-routes.

Before I do that though I thought I'd get some eyes on my methodology. If this is good and I can finish the grants route in this PR, then make another to switch organizations and materials over to the same pattern.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

Leaning on the unit tests heavily here.

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kurtwheeler** [comment]
Ok @davidsmejia and @Philinphiladelphia . This is no longer WIP! Please take a look. Sorry it got so big, but a lot of the size is because I squashed migrations.

File: api/resources_portal/test/factories.py [hunk 1]
```diff
@@ -86,3 +86,44 @@ def post(self, create, extracted, **kwargs):
...
+
+    organization = factory.SubFactory(PersonalOrganizationFactory)
+    grant = factory.SubFactory(LeafGrantFactory)
+
+
+class GrantMaterialAssociationFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = "resources_portal.GrantMaterialAssociation"
+
+    # Can I pass contact_user and organization along here?
```
Comment (Philinphiladelphia) [inline api/resources_portal/test/factories.py:119]: What do you mean here? Would you like contact_user and organization to be pulled out of your material factory? If you want to work with the SubFactory's data, you'll need to use the post generation hook, as seen here: https://factoryboy.readthedocs.io/en/v1.1.5/post_generation.html

File: api/resources_portal/test/views/test_grant.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        new_funder_id = "YYY-YYY-YYY"
+        grant_json["title"] = new_title
+        grant_json["funder_id"] = new_funder_id
+
+        # Test that users won't be updated.
+        new_member = UserFactory()
+        new_member_json = {"id": new_member.id}
+        grant_json["users"].append(new_member_json)
+
+        # TODO: this should require authentication
```
Comment (Philinphiladelphia) [inline api/resources_portal/test/views/test_grant.py:65]: You can add this in your code as-is. To authenticate during your tests, add

`self.client.force_authenticate(user=self.grant.users.first())`
in your test set-up. To check authentication on the ViewSet, override the create() method of your grant view and add the lines

`if not request.user.is_authenticated:`
`     return Response(status=status.HTTP_403_FORBIDDEN)`.

Or at least this is how I did it in [the invitations route](https://github.com/AlexsLemonade/resources-portal/pull/152). If there's a better way, let me know. If you end up doing this you'll also want to add a test for unauthenticated users being rejected.

File: api/resources_portal/urls.py [hunk 3]
```diff
@@ -6,7 +6,11 @@
 from rest_framework.authtoken import views
 from rest_framework.routers import DefaultRouter
 
-from resources_portal.views import (
+from rest_framework_extensions.routers import ExtendedSimpleRouter
+
+from resources_portal.views import (  # grant_material_relationship,; list_grant_material_relationships,
```
Comment (Philinphiladelphia) [inline api/resources_portal/urls.py:11]: Should this comment still be here?

File: api/resources_portal/views/relation_serializers.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+"""This file houses serializers which can be used for nested relationships.
+
+These serializers do not use nested relationships themselves, so that
+if a grant object links to a user and the user links to the grant, the
+JSON won't recur infinitely. For any relationships, these seializers
```
Comment (Philinphiladelphia) [inline api/resources_portal/views/relation_serializers.py:5]:
```suggestion
JSON won't recur infinitely. For any relationships, these serializers
```

File: api/resources_portal/test/factories.py (hunk 1 above)
Comment (kurtwheeler) [reply api/resources_portal/test/factories.py:119]: Sorry, I should have removed this already. I'm not exactly sure what I meant here, it was a note to myself from when I was first writing this.

File: api/resources_portal/test/views/test_grant_material_association.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        self.assertEqual(response.status_code, 201)
+        self.assertIn(material, self.grant.materials.all())
+
+    def test_post_fails_if_not_material_owner(self):
+        user = self.grant.users.first()
+        organization = self.grant.organizations.first()
+
+        self.client.force_authenticate(user=user)
+
+        # Organization's owener is a new user by default.
```
Comment (Philinphiladelphia) [inline api/resources_portal/test/views/test_grant_material_association.py:67]:
```suggestion
        # Organization's owner is a new user by default.
```

File: api/resources_portal/test/views/test_grant_material_association.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        url = reverse("grants-material-list", args=[self.grant.id])
+        response = self.client.post(url, data=model_to_dict(material))
+
+        self.assertEqual(response.status_code, 403)
+
+    def test_post_fails_if_not_grant_owner(self):
+        organization = self.grant.organizations.first()
+        self.client.force_authenticate(user=organization.owner)
+
+        # Organization's owener is a new user by default.
```
Comment (Philinphiladelphia) [inline api/resources_portal/test/views/test_grant_material_association.py:78]:
```suggestion
        # Organization's owner is a new user by default.
```

File: api/resources_portal/test/views/test_grant_material_association.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        url = reverse("grants-material-list", args=[self.grant.id])
+        response = self.client.post(url, data=model_to_dict(material))
+
+        self.assertEqual(response.status_code, 403)
+
+    def test_post_fails_if_grant_not_associated(self):
+        user = self.grant.users.first()
+        self.client.force_authenticate(user=user)
+
+        # Organization's owener is a new user by default.
```
Comment (Philinphiladelphia) [inline api/resources_portal/test/views/test_grant_material_association.py:89]:
```suggestion
        # Organization's owner is a new user by default.
```

File: api/resources_portal/test/views/test_grant_material_association.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        self.client.force_authenticate(user=user)
+        url = reverse("grants-material-list", args=[self.grant.id])
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 403)
+
+    def test_post_request_associates_a_material(self):
+        user = self.grant.users.first()
+        organization = self.grant.organizations.first()
+
+        # Organization's owener is a new user by default.
```
Comment (Philinphiladelphia) [inline api/resources_portal/test/views/test_grant_material_association.py:48]:
```suggestion
        # Organization's owner is a new user by default.
```

File: api/resources_portal/views/grant_material_association.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            request.user in grant.users.all()
+            and request.user == material.organization.owner
+            and grant in material.organization.grants.all()
+        )
+
+
+class GrantMaterialViewSet(NestedViewSetMixin, viewsets.ModelViewSet):
+    queryset = Material.objects.all().order_by("-created_at")
+
+    http_method_names = ["get", "post", "delete", "head"]
```
Comment (davidsmejia) [inline api/resources_portal/views/grant_material_association.py:38]: is options not required here?

Comment (kurtwheeler) [reply api/resources_portal/views/grant_material_association.py:38]: Oh good catch!

**Pull request merged**